packages/keyring-controller/src/KeyringController.ts

```diff
@@ -1874,13 +1906,97 @@ export class KeyringController<
       delete state.encryptionSalt;
     });
 
-    this.#password = password;
+    await this.#deriveEncryptionKey(password, {
+      ignoreExistingVault: true,
+    });
 
     await this.#clearKeyrings();
     await this.#createKeyringWithFirstAccount(keyring.type, keyring.opts);
     this.#setUnlocked();
   }
 
+  /**
+   * Derive the vault encryption key from the provided password, and
+   * assign it to the instance variable for later use with cryptographic
+   * functions.
+   *
+   * When the controller has a vault in its state, the key is derived
+   * using the salt from the vault. If the vault is empty, a new salt
+   * is generated and used to derive the key.
+   *
+   * If `options.ignoreExistingVault` is set to `true`, the existing
+   * vault is completely ignored: the new key won't be able to decrypt
+   * the existing vault, and should be used to re-encrypt it.
+   *
+   * @param password - The password to use for decryption or derivation.
+   * @param options - Options for the key derivation.
+   * @param options.ignoreExistingVault - Whether to ignore the existing vault salt and key metadata
+   */
+  async #deriveEncryptionKey(
```

Comment:
nit
```suggestion
  async #deriveAndSetEncryptionKey(
```
(to emphasize that we also set the key here and `#setEncryptionKey` doesn't need to be called afterwards)

Reply:
Good point! renamed in 9d1bb4d18a61